pipit/readers/otf2_reader.py

```diff
@@ -190,70 +192,130 @@ def events_reader(self, rank_size):
             # columns of the DataFrame
             timestamps, event_types, event_attributes, names = [], [], [], []
 
-            # Note:
-            # 1. The below lists are for storing logical ids.
-            # 2. Support for GPU events has to be added and unified across readers.
+            # note: the below lists are for storing logical ids
             process_ids, thread_ids = [], []
 
+            # list of metric members of in the definitions
+            metric_names = set(
+                self.definitions.loc[
+                    self.definitions["Definition Type"] == "MetricMember"
+                ]["Attributes"]
+                .map(lambda attr: attr["name"])
+                .to_list()
+            )
+
+            # maps each metric to a list of its values
+            metrics_dict = {metric_name: [] for metric_name in metric_names}
+
+            # used to keep track of information about
+            # the most recent metrics that were read
+            prev_metric_time = -1
+            prev_metrics = set()
+
             # iterates through the events and processes them
             for loc_event in loc_events:
                 # extracts the location and event
                 # location could be thread, process, etc
                 loc, event = loc_event[0], loc_event[1]
 
-                # information about the location that the event occurred on
-
-                # TO DO:
-                # need to add support for accelerator and metric locations
+                # To Do:
+                # Support for GPU events has to be
+                # added and unified across readers.
                 if str(loc.type)[13:] == "CPU_THREAD":
-                    process_id = loc.group._ref
-                    process_ids.append(process_id)
-
-                    # subtract the minimum location number of a process
-                    # from the location number to get threads numbered
-                    # 0 to (num_threads per process - 1) for each process.
-                    thread_ids.append(loc._ref - self.process_threads_map[process_id])
-
-                    # type of event - enter, leave, or other types
-                    event_type = str(type(event))[20:-2]
-                    if event_type == "Enter" or event_type == "Leave":
-                        event_types.append(event_type)
-                    else:
-                        event_types.append("Instant")
+                    # don't add metric events as a separate row,
+                    # and add their values into columns instead
+                    if type(event) == otf2.events.Metric:
+                        # Since the location is a cpu thread, we know
+                        # that the metric event is of type MetricClass,
+                        # which has a list of MetricMembers.
+                        metrics = list(
```

Comment:
I think I had this as a check because there were some metric members which were in the definitions, not in the trace. But you are right. I think I can fix this by setting metric_names to be all the metrics that appear in metric classes, because those are actually what show up in the trace.